### PR TITLE
 add enclave upload impl

### LINKED ISSUE #23: Add a message handler for the DataEnclaveActor to handle the data payload
Rename Enclave actor to DataEnclave actor

 - [x] Pass the payload to the enclave and wait for the response
 - [x] Pass the sealed data from the Enclave's Ocall to the Azure Blob storage handling code to be saved there. (Just save to disk now, and create interface / module for saving files)
 - [x] Return the payload that the enclave encrypted for the client.

### PR DESCRIPTION
**Pull request by longtomjr**

### CODE REVIEW

File: rtc_data_enclave/Cargo.toml [hunk 1]
```diff
@@ -32,6 +35,16 @@ bincode = { git = "https://github.com/mesalock-linux/bincode-sgx.git" }
...
 simple_asn1 = { git = "https://github.com/mesalock-linux/simple_asn1-sgx.git" }
 thiserror = { git = "https://github.com/mesalock-linux/thiserror-sgx.git" }
-
+uuid = { git = "https://github.com/mesalock-linux/uuid-sgx", features = ["v4"] }
+rand = { git = "https://github.com/mesalock-linux/rand-sgx", tag = "v0.7.3_sgx1.1.3" }
 # No Std Dependencies
 zeroize = "1.2.0"
+secrecy = { version = "0.7.0", default-features = false }
+# Using the git version here for no-std support
+sodiumoxide = { git = "https://github.com/sodiumoxide/sodiumoxide.git", default-features = false }
```
Comment (longtomjr) [inline rtc_data_enclave/Cargo.toml]: I don't think we are using this anymore.

File: rtc_data_enclave/src/data_upload.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+pub struct SealedResult {
+    /// Uploaded data sealed for this enclave
+    pub sealed_data: Box<[u8]>,
+    /// Payload for client encrypted with the client's ephemeral key
+    pub client_payload: DataUploadResponse,
+
+    pub uuid: Uuid,
+}
+
+pub struct UploadedData {
```
Comment (longtomjr) [inline rtc_data_enclave/src/data_upload.rs]: Check if this is used and remove if unused

File: rtc_data_enclave/src/data_upload.rs [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    ephemeral_key: Box<[u8]>,
+    data: Box<[u8]>,
+}
+
+pub fn validate_and_seal(payload: UploadPayload) -> Result<SealedResult, DataError> {
```
Comment (longtomjr) [inline rtc_data_enclave/src/data_upload.rs]: Pass in 2 args instead of using struct declaration

File: rtc_data_enclave/src/crypto.rs [hunk 4]
```diff
@@ -138,7 +162,7 @@ impl RtcCrypto for SodaBoxCrypto {
         // be applicable to all situations
         match self.rng.try_fill(&mut nonce) {
             Ok(_) => Ok(nonce),
-            Err(err) => Err(self::Error::Rand(err)),
+            Err(err) => Err(self::Error::Rand(err.code().map_or(0, |code| code.get()))),
```
Comment (PiDelport) [inline rtc_data_enclave/src/crypto.rs:166]: May be worth adding a comment here?

File: rtc_data_enclave/src/data_upload.rs [hunk 5]
```diff
@@ -76,21 +60,19 @@ pub fn validate_and_seal(payload: UploadPayload) -> Result<SealedResult, DataErr
 fn generate_client_payload(
     their_pk: &[u8; 32],
     crypto: &mut Crypto,
-) -> Result<(EncryptedMessage, Uuid), CryptoError> {
+) -> Result<(SizedEncryptedMessage<{ 24 + 16 + 32 }>, Uuid), CryptoError> {
```
Comment (PiDelport) [inline rtc_data_enclave/src/data_upload.rs]: Use `DataUploadResponse` here?

File: rtc_types/src/ecall_result.rs [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+impl<T, E> Into<Result<T, E>> for EcallResult<T, E> {
+    fn into(self) -> Result<T, E> {
+        match self {
+            EcallResult::Ok(res) => Result::Ok(res),
+            EcallResult::Err(err) => Result::Err(err),
+        }
+    }
+}
+
```
Comment (PiDelport) [inline rtc_types/src/ecall_result.rs]: We should probably add a comment here for why it's `Into` instead of `From`?

**Review by PiDelport — COMMENTED** [review]
Capture comments from yesterday's session with @longtomjr.

**Comment by PiDelport** [comment]
(Rebased to resolve trivial conflict.)

**Review by PiDelport — APPROVED** [review]

File: rtc_data_enclave/src/data_upload.rs (hunk 3 above)
Comment (Nghondzweni) [reply rtc_data_enclave/src/data_upload.rs]: @longtomjr have you checked and made these changes we discussed in the review yesterday?

File: rtc_types/src/ecall_result.rs (hunk 6 above)
Comment (PiDelport) [reply rtc_types/src/ecall_result.rs]: It looks like implementing `From` works, so I'll do that.

**Comment by PiDelport** [comment]
@longtomjr, @Nghondzweni: Cool, I addressed the feedback points. This should be ready to merge once you've looked over it?

**Review by Nghondzweni — APPROVED** [review]

**Pull request merged**